Sources/NIOExtras/HTTP1ProxyConnectHandler.swift

```
@@ -0,0 +1,393 @@
+        switch self.state {
+        case .initialized, .connectSent, .headReceived, .completed:
+            self.bufferedWrittenMessages.append(BufferedWrite(data: data, promise: promise))
+        case .failed(let error):
+            promise?.fail(error)
+        }
+    }
+
+    public func flush(context: ChannelHandlerContext) {
+        self.bufferedWrittenMessages.mark()
+    }
+
+    public func removeHandler(context: ChannelHandlerContext, removalToken: ChannelHandlerContext.RemovalToken) {
+        // We have been formally removed from the pipeline. We should send any buffered data we have.
+        switch self.state {
+        case .initialized, .connectSent, .headReceived, .failed:
+            self.failWithError(.noResult(), context: context)
+
+        case .completed:
+            let hadMark = self.bufferedWrittenMessages.hasMark
+            while self.bufferedWrittenMessages.hasMark {
+                // write until mark
+                let bufferedPart = self.bufferedWrittenMessages.removeFirst()
+                context.write(bufferedPart.data, promise: bufferedPart.promise)
+            }
+
+            // flush any messages up to the mark
+            if hadMark {
+                context.flush()
+            }
+
+            // write remainder
+            while let bufferedPart = self.bufferedWrittenMessages.popFirst() {
+                context.write(bufferedPart.data, promise: bufferedPart.promise)
+            }
+        }
+
+        context.leavePipeline(removalToken: removalToken)
+    }
+
+    public func handlerAdded(context: ChannelHandlerContext) {
+        if context.channel.isActive {
+            self.sendConnect(context: context)
+        }
+    }
+
+    public func handlerRemoved(context: ChannelHandlerContext) {
+        switch self.state {
+        case .failed, .completed:
+            // we don't expect there to be any buffered messages in these states
+            assert(self.bufferedWrittenMessages.isEmpty)
```

This case shouldn't pass silently: we should either throw an error, precondition, or handle the edge-case.